gen-guest-rust: cannot reuse shared records between import and export interfaces

The following wit file:

```
interface types {
    record a {
        b: string    
    }
}

interface foo {
    use self.types.{a}

    x: func(param: a)
}

interface bar {
    use self.types.{a}

    x: func(param: a)
    y: func() -> a
}

default world demo {
    import foo: self.foo
    export bar: self.bar
}
```

Fails to compile with the following error:

```rust
error[E0422]: cannot find struct, variant or union type `AParam` in this scope
 --> src/lib.rs:3:1
  |
3 | / wit_bindgen::generate!({
4 | |     world: "demo",
5 | | });
  | |__^ not found in this scope
  |
  = help: consider importing this struct:
          crate::types::AParam
  = note: this error originates in the macro `wit_bindgen::generate` (in Nightly builds, run with -Z macro-backtrace for more info)
error[E0412]: cannot find type `A` in module `super::types`
 --> src/lib.rs:3:1
  |
3 | / wit_bindgen::generate!({
4 | |     world: "demo",
5 | | });
  | |__^ not found in `super::types`
  |
  = help: consider importing this type alias:
          crate::foo::A
  = note: this error originates in the macro `wit_bindgen::generate` (in Nightly builds, run with -Z macro-backtrace for more info)
  ```
My findings:

The generated Rust type for `record a` gets split into two types and looks like this:

```rust
pub mod types {
    pub struct AParam<'a> {
        pub b: &'a str,
    }
    ...
    pub struct AResult {
        pub b: wit_bindgen::rt::string::String,
    }
    ...
}
```

However, The `bar` module attempts to import `record a` as `super::types::A;`:

```rust
pub mod bar {
    pub type A = super::types::A;
    ...
}
```

Conditions to encounter this error:

- one `import` interface uses a record (with a non-primitive field) as a function parameter
- an `export` interface uses the same record as either parameter or return type

Tested with:

- v0.3.0
- commit: 462527127f0d9de935f01e298a7e1c975e124a62

--- a/crates/bindgen-core/src/lib.rs
+++ b/crates/bindgen-core/src/lib.rs
@@ -16,12 +16,18 @@ pub struct Types {
 #[derive(Default, Clone, Copy, Debug)]
 pub struct TypeInfo {
     /// Whether or not this type is ever used (transitively) within the
-    /// parameter of a function.
-    pub param: bool,
+    /// parameter of an imported function.
+    ///
+    /// This means that it's used in a context where ownership isn't
+    /// relinquished.
+    pub borrowed: bool,
 
     /// Whether or not this type is ever used (transitively) within the
-    /// result of a function.
-    pub result: bool,
+    /// parameter or result of an export, or the result of an import.
+    ///
+    /// This means that it's used in a context where ownership is required and
+    /// memory management is necessary.
+    pub owned: bool,
 
     /// Whether or not this type is ever used (transitively) within the
     /// error case in the result of a function.
@@ -33,8 +39,8 @@ pub struct TypeInfo {
 
 impl std::ops::BitOrAssign for TypeInfo {
     fn bitor_assign(&mut self, rhs: Self) {
-        self.param |= rhs.param;
-        self.result |= rhs.result;
+        self.borrowed |= rhs.borrowed;
+        self.owned |= rhs.owned;
         self.error |= rhs.error;
         self.has_list |= rhs.has_list;
     }
@@ -45,29 +51,63 @@ impl Types {
         for (t, _) in resolve.types.iter() {
             self.type_id_info(resolve, t);
         }
-        for (_, iface) in resolve.interfaces.iter() {
-            for (_, f) in iface.functions.iter() {
-                self.type_info_func(resolve, f);
-            }
-        }
         for (_, world) in resolve.worlds.iter() {
-            for (_, item) in world.imports.iter().chain(&world.exports) {
+            for (import, (_, item)) in world
+                .imports
+                .iter()
+                .map(|i| (true, i))
+                .chain(world.exports.iter().map(|i| (false, i)))
+            {
                 match item {
                     WorldItem::Function(f) => {
-                        self.type_info_func(resolve, f);
+                        self.type_info_func(resolve, f, import);
                     }
-                    WorldItem::Interface(_) | WorldItem::Type(_) => {}
+                    WorldItem::Interface(id) => {
+                        for (_, f) in resolve.interfaces[*id].functions.iter() {
+                            self.type_info_func(resolve, f, import);
+                        }
+                    }
+                    WorldItem::Type(_) => {}
                 }
             }
         }
     }
 
-    fn type_info_func(&mut self, resolve: &Resolve, f: &Function) {
-        for (_, ty) in f.params.iter() {
-            self.set_param_result_ty(resolve, ty, true, false, false);
+    fn type_info_func(&mut self, resolve: &Resolve, func: &Function, import: bool) {
+        let mut live = LiveTypes::default();
+        for (_, ty) in func.params.iter() {
+            self.type_info(resolve, ty);
+            live.add_type(resolve, ty);
         }
-        for ty in f.results.iter_types() {
-            self.set_param_result_ty(resolve, ty, false, true, false);
+        for id in live.iter() {
+            let info = self.type_info.get_mut(&id).unwrap();
+            if import {
+                info.borrowed = true;
+            } else {
+                info.owned = true;
+            }
+        }
+        let mut live = LiveTypes::default();
+        for ty in func.results.iter_types() {
+            self.type_info(resolve, ty);
+            live.add_type(resolve, ty);
+        }
+        for id in live.iter() {
+            self.type_info.get_mut(&id).unwrap().owned = true;
+        }
+
+        for ty in func.results.iter_types() {
+            let id = match ty {
+                Type::Id(id) => *id,
+                _ => continue,
+            };
+            let err = match &resolve.types[id].kind {
+                TypeDefKind::Result(Result_ { err, .. }) => err,
+                _ => continue,
+            };
+            if let Some(Type::Id(id)) = err {
+                self.type_info.get_mut(&id).unwrap().error = true;
+            }
         }
     }
 
@@ -144,110 +184,6 @@ impl Types {
         match ty {
             Some(ty) => self.type_info(resolve, ty),
             None => TypeInfo::default(),
-        }
-    }
-
-    fn set_param_result_id(
-        &mut self,
-        resolve: &Resolve,
-        ty: TypeId,
-        param: bool,
-        result: bool,
-        error: bool,
-    ) {
-        match &resolve.types[ty].kind {
-            TypeDefKind::Record(r) => {
-                for field in r.fields.iter() {
-                    self.set_param_result_ty(resolve, &field.ty, param, result, error)
-                }
-            }
-            TypeDefKind::Tuple(t) => {
-                for ty in t.types.iter() {
-                    self.set_param_result_ty(resolve, ty, param, result, error)
-                }
-            }
-            TypeDefKind::Flags(_) => {}
-            TypeDefKind::Enum(_) => {}
-            TypeDefKind::Variant(v) => {
-                for case in v.cases.iter() {
-                    self.set_param_result_optional_ty(
-                        resolve,
-                        case.ty.as_ref(),
-                        param,
-                        result,
-                        error,
-                    )
-                }
-            }
-            TypeDefKind::List(ty) | TypeDefKind::Type(ty) | TypeDefKind::Option(ty) => {
-                self.set_param_result_ty(resolve, ty, param, result, error)
-            }
-            TypeDefKind::Result(r) => {
-                self.set_param_result_optional_ty(resolve, r.ok.as_ref(), param, result, error);
-                self.set_param_result_optional_ty(resolve, r.err.as_ref(), param, result, result);
-            }
-            TypeDefKind::Union(u) => {
-                for case in u.cases.iter() {
-                    self.set_param_result_ty(resolve, &case.ty, param, result, error)
-                }
-            }
-            TypeDefKind::Future(ty) => {
-                self.set_param_result_optional_ty(resolve, ty.as_ref(), param, result, error)
-            }
-            TypeDefKind::Stream(stream) => {
-                self.set_param_result_optional_ty(
-                    resolve,
-                    stream.element.as_ref(),
-                    param,
-                    result,
-                    error,
-                );
-                self.set_param_result_optional_ty(
-                    resolve,
-                    stream.end.as_ref(),
-                    param,
-                    result,
-                    error,
-                );
-            }
-            TypeDefKind::Unknown => unreachable!(),
-        }
-    }
-
-    fn set_param_result_ty(
-        &mut self,
-        resolve: &Resolve,
-        ty: &Type,
-        param: bool,
-        result: bool,
-        error: bool,
-    ) {
-        match ty {
-            Type::Id(id) => {
-                self.type_id_info(resolve, *id);
-                let info = self.type_info.get_mut(id).unwrap();
-                if (param && !info.param) || (result && !info.result) || (error && !info.error) {
-                    info.param = info.param || param;
-                    info.result = info.result || result;
-                    info.error = info.error || error;
-                    self.set_param_result_id(resolve, *id, param, result, error);
-                }
-            }
-            _ => {}
-        }
-    }
-
-    fn set_param_result_optional_ty(
-        &mut self,
-        resolve: &Resolve,
-        ty: Option<&Type>,
-        param: bool,
-        result: bool,
-        error: bool,
-    ) {
-        match ty {
-            Some(ty) => self.set_param_result_ty(resolve, ty, param, result, error),
-            None => (),
         }
     }
 }

--- a/crates/gen-guest-c/src/lib.rs
+++ b/crates/gen-guest-c/src/lib.rs
@@ -1115,21 +1115,7 @@ impl InterfaceGenerator<'_> {
                 let ty = &self.resolve.types[*id];
                 match &ty.name {
                     Some(name) => {
-                        match ty.owner {
-                            TypeOwner::Interface(owner) => {
-                                self.src.print(
-                                    stype,
-                                    &self.gen.interface_names[&owner].to_snake_case(),
-                                );
-                                self.src.print(stype, "_");
-                            }
-                            TypeOwner::World(owner) => {
-                                self.src
-                                    .print(stype, &self.resolve.worlds[owner].name.to_snake_case());
-                                self.src.print(stype, "_");
-                            }
-                            TypeOwner::None => {}
-                        }
+                        self.print_owner_namespace(stype, *id);
 
                         self.src.print(stype, &name.to_snake_case());
                         self.src.print(stype, "_t");
@@ -1145,6 +1131,25 @@ impl InterfaceGenerator<'_> {
                         }
                     },
                 }
+            }
+        }
+    }
+
+    fn print_owner_namespace(&mut self, stype: SourceType, id: TypeId) {
+        let ty = &self.resolve.types[id];
+        match ty.owner {
+            TypeOwner::Interface(owner) => {
+                self.src
+                    .print(stype, &self.gen.interface_names[&owner].to_snake_case());
+                self.src.print(stype, "_");
+            }
+            TypeOwner::World(owner) => {
+                self.src
+                    .print(stype, &self.resolve.worlds[owner].name.to_snake_case());
+                self.src.print(stype, "_");
+            }
+            TypeOwner::None => {
+                self.print_namespace(stype);
             }
         }
     }
@@ -1456,6 +1461,9 @@ impl InterfaceGenerator<'_> {
             Type::String => {
                 self.src.h_helpers(&self.gen.world.to_snake_case());
                 self.src.h_helpers("_");
+            }
+            Type::Id(id) => {
+                self.print_owner_namespace(SourceType::HHelpers, *id);
             }
             _ => {
                 self.print_namespace(SourceType::HHelpers);

--- a/crates/gen-guest-rust/src/lib.rs
+++ b/crates/gen-guest-rust/src/lib.rs
@@ -86,7 +86,6 @@ impl RustWasm {
         &'a mut self,
         wasm_import_module: Option<&'a str>,
         resolve: &'a Resolve,
-        default_param_mode: TypeMode,
         in_import: bool,
     ) -> InterfaceGenerator<'a> {
         let mut sizes = SizeAlign::default();
@@ -100,7 +99,6 @@ impl RustWasm {
             gen: self,
             sizes,
             resolve,
-            default_param_mode,
             return_pointer_area_size: 0,
             return_pointer_area_align: 0,
         }
@@ -126,7 +124,7 @@ impl WorldGenerator for RustWasm {
     ) {
         let prev = self.interface_names.insert(id, name.to_snake_case());
         assert!(prev.is_none());
-        let mut gen = self.interface(Some(name), resolve, TypeMode::AllBorrowed("'a"), true);
+        let mut gen = self.interface(Some(name), resolve, true);
         gen.current_interface = Some(id);
         gen.types(id);
 
@@ -144,7 +142,7 @@ impl WorldGenerator for RustWasm {
         funcs: &[(&str, &Function)],
         _files: &mut Files,
     ) {
-        let mut gen = self.interface(Some("$root"), resolve, TypeMode::AllBorrowed("'a"), true);
+        let mut gen = self.interface(Some("$root"), resolve, true);
 
         for (_, func) in funcs {
             gen.generate_guest_import(func);
@@ -162,7 +160,7 @@ impl WorldGenerator for RustWasm {
         _files: &mut Files,
     ) {
         self.interface_names.insert(id, name.to_snake_case());
-        let mut gen = self.interface(None, resolve, TypeMode::Owned, false);
+        let mut gen = self.interface(None, resolve, false);
         gen.current_interface = Some(id);
         gen.types(id);
         gen.generate_exports(name, Some(name), resolve.interfaces[id].functions.values());
@@ -177,7 +175,7 @@ impl WorldGenerator for RustWasm {
         _files: &mut Files,
     ) {
         let name = &resolve.worlds[world].name;
-        let mut gen = self.interface(None, resolve, TypeMode::Owned, false);
+        let mut gen = self.interface(None, resolve, false);
         gen.generate_exports(name, None, funcs.iter().map(|f| f.1));
         let src = gen.finish();
         self.src.push_str(&src);
@@ -190,7 +188,7 @@ impl WorldGenerator for RustWasm {
         types: &[(&str, TypeId)],
         _files: &mut Files,
     ) {
-        let mut gen = self.interface(None, resolve, TypeMode::Owned, false);
+        let mut gen = self.interface(None, resolve, false);
         for (name, ty) in types {
             gen.define_type(name, *ty);
         }
@@ -321,7 +319,6 @@ struct InterfaceGenerator<'a> {
     gen: &'a mut RustWasm,
     wasm_import_module: Option<&'a str>,
     resolve: &'a Resolve,
-    default_param_mode: TypeMode,
     return_pointer_area_size: usize,
     return_pointer_area_align: usize,
 }
@@ -632,10 +629,6 @@ impl<'a> RustGenerator<'a> for InterfaceGenerator<'a> {
 
     fn string_name(&self) -> &'static str {
         "wit_bindgen::rt::string::String"
-    }
-
-    fn default_param_mode(&self) -> TypeMode {
-        self.default_param_mode
     }
 
     fn push_str(&mut self, s: &str) {

--- a/crates/gen-rust-lib/src/lib.rs
+++ b/crates/gen-rust-lib/src/lib.rs
@@ -38,7 +38,6 @@ pub trait RustGenerator<'a> {
     fn types_mut(&mut self) -> &mut Types;
     fn print_borrowed_slice(&mut self, mutbl: bool, ty: &Type, lifetime: &'static str);
     fn print_borrowed_str(&mut self, lifetime: &'static str);
-    fn default_param_mode(&self) -> TypeMode;
 
     fn rustdoc(&mut self, docs: &Docs) {
         let docs = match &docs.contents {
@@ -207,11 +206,11 @@ pub trait RustGenerator<'a> {
         }
     }
 
-    fn type_path(&self, id: TypeId, param: bool) -> String {
-        let name = if param {
-            self.param_name(id)
-        } else {
+    fn type_path(&self, id: TypeId, owned: bool) -> String {
+        let name = if owned {
             self.result_name(id)
+        } else {
+            self.param_name(id)
         };
         if let TypeOwner::Interface(id) = self.resolve().types[id].owner {
             if let Some(path) = self.path_to_interface(id) {
@@ -226,13 +225,13 @@ pub trait RustGenerator<'a> {
         let lt = self.lifetime_for(&info, mode);
         let ty = &self.resolve().types[id];
         if ty.name.is_some() {
-            let name = self.type_path(id, lt.is_some());
+            let name = self.type_path(id, lt.is_none());
             self.push_str(&name);
 
             // If the type recursively owns data and it's a
             // variant/record/list, then we need to place the
             // lifetime parameter on the type as well.
-            if info.owns_data() && needs_generics(self.resolve(), &ty.kind) {
+            if info.has_list && needs_generics(self.resolve(), &ty.kind) {
                 self.print_generics(lt);
             }
 
@@ -383,11 +382,9 @@ pub trait RustGenerator<'a> {
     fn modes_of(&self, ty: TypeId) -> Vec<(String, TypeMode)> {
         let info = self.info(ty);
         let mut result = Vec::new();
-        if info.param {
-            result.push((self.param_name(ty), self.default_param_mode()));
-        }
-        if info.result && (!info.param || self.uses_two_names(&info)) {
-            result.push((self.result_name(ty), TypeMode::Owned));
+        result.push((self.result_name(ty), TypeMode::Owned));
+        if self.uses_two_names(&info) {
+            result.push((self.param_name(ty), TypeMode::AllBorrowed("'a")));
         }
         return result;
     }
@@ -524,7 +521,7 @@ pub trait RustGenerator<'a> {
                 self.push_str("#[component(record)]\n");
             }
 
-            if !info.owns_data() {
+            if !info.has_list {
                 self.push_str("#[repr(C)]\n");
                 self.push_str("#[derive(Copy, Clone)]\n");
             } else {
@@ -680,7 +677,7 @@ pub trait RustGenerator<'a> {
                 self.push_str("#[derive(wasmtime::component::Lower)]\n");
                 self.push_str(&format!("#[component({})]\n", derive_component));
             }
-            if !info.owns_data() {
+            if !info.has_list {
                 self.push_str("#[derive(Clone, Copy)]\n");
             } else {
                 self.push_str("#[derive(Clone)]\n");
@@ -977,13 +974,7 @@ pub trait RustGenerator<'a> {
     }
 
     fn uses_two_names(&self, info: &TypeInfo) -> bool {
-        info.owns_data()
-            && info.param
-            && info.result
-            && match self.default_param_mode() {
-                TypeMode::AllBorrowed(_) | TypeMode::LeafBorrowed(_) => true,
-                TypeMode::Owned => false,
-            }
+        info.has_list && info.borrowed && info.owned
     }
 
     fn lifetime_for(&self, info: &TypeInfo, mode: TypeMode) -> Option<&'static str> {
@@ -1104,19 +1095,15 @@ pub trait RustFunctionGenerator {
     }
 
     fn typename_lower(&self, id: TypeId) -> String {
-        let param = match self.lift_lower() {
-            LiftLower::LowerArgsLiftResults => true,
-            LiftLower::LiftArgsLowerResults => false,
+        let owned = match self.lift_lower() {
+            LiftLower::LowerArgsLiftResults => false,
+            LiftLower::LiftArgsLowerResults => true,
         };
-        self.rust_gen().type_path(id, param)
+        self.rust_gen().type_path(id, owned)
     }
 
     fn typename_lift(&self, id: TypeId) -> String {
-        let param = match self.lift_lower() {
-            LiftLower::LiftArgsLowerResults => true,
-            LiftLower::LowerArgsLiftResults => false,
-        };
-        self.rust_gen().type_path(id, param)
+        self.rust_gen().type_path(id, true)
     }
 }
 
@@ -1193,16 +1180,6 @@ pub fn int_repr(repr: Int) -> &'static str {
         Int::U16 => "u16",
         Int::U32 => "u32",
         Int::U64 => "u64",
-    }
-}
-
-trait TypeInfoExt {
-    fn owns_data(&self) -> bool;
-}
-
-impl TypeInfoExt for TypeInfo {
-    fn owns_data(&self) -> bool {
-        self.has_list
     }
 }
 

--- a/tests/codegen/lift-lower-foreign.wit
+++ b/tests/codegen/lift-lower-foreign.wit
@@ -8,11 +8,14 @@ interface a {
   type t7 = option<t2>
   type t8 = result<t2, t3>
   union t9 { t1, t2, t3, t4, u32 }
+  type t10 = list<t5>
+  type t11 = t10
 }
 
 interface the-interface {
   use self.a.{t1 as u1, t2 as u2, t3 as u3, t4 as u4, t5 as u5}
   use self.a.{t6 as u6, t7 as u7, t8 as %u8, t9 as u9}
+  use self.a.{t10 as u10, t11 as u11}
 
   f1: func(a: u1) -> u1
   f2: func(a: u2) -> u2
@@ -23,11 +26,14 @@ interface the-interface {
   f7: func(a: u7) -> u7
   f8: func(a: %u8) -> %u8
   f9: func(a: u9) -> u9
+  f10: func(a: u10) -> u10
+  f11: func(a: u11) -> u11
 }
 
 default world foo {
   use self.a.{t1 as u1, t2 as u2, t3 as u3, t4 as u4, t5 as u5}
   use self.a.{t6 as u6, t7 as u7, t8 as %u8, t9 as u9}
+  use self.a.{t10 as u10, t11 as u11}
 
   export f1: func(a: u1) -> u1
   export f2: func(a: u2) -> u2
@@ -38,6 +44,8 @@ default world foo {
   export f7: func(a: u7) -> u7
   export f8: func(a: %u8) -> %u8
   export f9: func(a: u9) -> u9
+  export f10: func(a: u10) -> u10
+  export f11: func(a: u11) -> u11
 
   import the-import: self.the-interface
   export the-export: self.the-interface

--- a/tests/runtime/flavorful/wasm.rs
+++ b/tests/runtime/flavorful/wasm.rs
@@ -12,7 +12,7 @@ impl Flavorful for Component {
 
         let _guard = test_rust_wasm::guard();
 
-        f_list_in_record1(ListInRecord1 {
+        f_list_in_record1(ListInRecord1Param {
             a: "list_in_record1",
         });
         assert_eq!(f_list_in_record2().a, "list_in_record2");
@@ -30,7 +30,11 @@ impl Flavorful for Component {
             "result4"
         );
 
-        f_list_in_variant1(Some("foo"), Err("bar"), ListInVariant1V3::String("baz"));
+        f_list_in_variant1(
+            Some("foo"),
+            Err("bar"),
+            ListInVariant1V3Param::String("baz"),
+        );
         assert_eq!(f_list_in_variant2(), Some("list_in_variant2".to_string()));
         assert_eq!(
             f_list_in_variant3(Some("input3")),
@@ -62,7 +66,7 @@ impl Flavorful for Component {
 }
 
 impl exports::Exports for Component {
-    fn f_list_in_record1(ty: ListInRecord1) {
+    fn f_list_in_record1(ty: ListInRecord1Result) {
         assert_eq!(ty.a, "list_in_record1");
     }
 
@@ -72,26 +76,30 @@ impl exports::Exports for Component {
         }
     }
 
-    fn f_list_in_record3(a: ListInRecord3) -> ListInRecord3 {
+    fn f_list_in_record3(a: ListInRecord3Result) -> ListInRecord3Result {
         assert_eq!(a.a, "list_in_record3 input");
-        ListInRecord3 {
+        ListInRecord3Result {
             a: "list_in_record3 output".to_string(),
         }
     }
 
-    fn f_list_in_record4(a: ListInAlias) -> ListInAlias {
+    fn f_list_in_record4(a: ListInAliasResult) -> ListInAliasResult {
         assert_eq!(a.a, "input4");
-        ListInRecord4 {
+        ListInRecord4Result {
             a: "result4".to_string(),
         }
     }
 
-    fn f_list_in_variant1(a: ListInVariant1V1, b: ListInVariant1V2, c: ListInVariant1V3) {
+    fn f_list_in_variant1(
+        a: ListInVariant1V1Result,
+        b: ListInVariant1V2Result,
+        c: ListInVariant1V3Result,
+    ) {
         assert_eq!(a.unwrap(), "foo");
         assert_eq!(b.unwrap_err(), "bar");
         match c {
-            ListInVariant1V3::String(s) => assert_eq!(s, "baz"),
-            ListInVariant1V3::F32(_) => panic!(),
+            ListInVariant1V3Result::String(s) => assert_eq!(s, "baz"),
+            ListInVariant1V3Result::F32(_) => panic!(),
         }
     }
 
@@ -99,7 +107,7 @@ impl exports::Exports for Component {
         Some("list_in_variant2".to_string())
     }
 
-    fn f_list_in_variant3(a: ListInVariant3) -> Option<String> {
+    fn f_list_in_variant3(a: ListInVariant3Result) -> Option<String> {
         assert_eq!(a.unwrap(), "input3");
         Some("output3".to_string())
     }
@@ -112,7 +120,10 @@ impl exports::Exports for Component {
         Err(MyErrno::B)
     }
 
-    fn list_typedefs(a: ListTypedef, b: ListTypedef3) -> (ListTypedef2, ListTypedef3) {
+    fn list_typedefs(
+        a: ListTypedefResult,
+        b: ListTypedef3Result,
+    ) -> (ListTypedef2, ListTypedef3Result) {
         assert_eq!(a, "typedef1");
         assert_eq!(b.len(), 1);
         assert_eq!(b[0], "typedef2");

--- a/tests/runtime/unions/wasm.rs
+++ b/tests/runtime/unions/wasm.rs
@@ -188,10 +188,10 @@ impl exports::Exports for Component {
         }
     }
 
-    fn replace_first_char(text: AllText, letter: char) -> AllText {
+    fn replace_first_char(text: AllTextResult, letter: char) -> AllTextResult {
         match text {
-            AllText::Char(_c) => AllText::Char(letter),
-            AllText::String(s) => AllText::String(format!("{}{}", letter, &s[1..])),
+            AllTextResult::Char(_c) => AllTextResult::Char(letter),
+            AllTextResult::String(s) => AllTextResult::String(format!("{}{}", letter, &s[1..])),
         }
     }
 
@@ -219,10 +219,10 @@ impl exports::Exports for Component {
         }
     }
 
-    fn identify_text(text: AllText) -> u8 {
+    fn identify_text(text: AllTextResult) -> u8 {
         match text {
-            AllText::Char(_c) => 0,
-            AllText::String(_s) => 1,
+            AllTextResult::Char(_c) => 0,
+            AllTextResult::String(_s) => 1,
         }
     }
 

